tests/test_validators.py

```diff
@@ -1,332 +1,389 @@
 
-        # When
+    # Then raises
+    with pytest.raises(validators.Invalid):
         unique_validator('1')
-        unique_validator('2')
-        unique_validator('3')
-        unique_validator('foo')
 
-        # Then raises
-        with pytest.raises(validators.Invalid):
-            unique_validator('1')
 
-    def test_unique_validators_do_not_cross_validate(self):
-        # Given
-        unique_validator_1 = validators.unique()
-        unique_validator_2 = validators.unique()
+def test_unique_validators_do_not_cross_validate():
+    # Given
+    unique_validator_1 = validators.unique()
+    unique_validator_2 = validators.unique()
+
+    # When
+    unique_validator_1('a')
+    unique_validator_2('a')
+
+    # Then no invalid exception is raised
+
+
+def test_mandatory_valid():
+    # Given
+    mandatory_validator = validators.mandatory()
+
+    # When
+    mandatory_validator('a')
+
+    # Then no invalid exception is raised
+
+
+def test_mandatory_invalid():
+    # Given
+    mandatory_validator = validators.mandatory()
+
+    # When, then raises
+    with pytest.raises(validators.Invalid):
+        mandatory_validator('')
+
+
+def test_numeric_valid():
+    # Given
+    numeric_validator = validators.numeric()
+
+    # When
+    numeric_validator('0123456789')
+
+    # Then no invalid exception is raised
+
+
+def test_numeric_invalid():
+    # Given
+    numeric_validator = validators.numeric()
 
-        # When
-        unique_validator_1('a')
-        unique_validator_2('a')
+    # When, then raises
+    with pytest.raises(validators.Invalid):
+        numeric_validator('a')
 
-        # Then no invalid exception is raised
+    with pytest.raises(validators.Invalid):
+        numeric_validator('1.1')
 
-    def test_mandatory_valid(self):
-        # Given
-        mandatory_validator = validators.mandatory()
+    with pytest.raises(validators.Invalid):
+        numeric_validator('_')
 
-        # When
-        mandatory_validator('a')
 
-        # Then no invalid exception is raised
+def test_lat_long_valid():
+    # Given
+    lat_long_validator = validators.latitude_longitude(max_scale=5, max_precision=10)
 
-    def test_mandatory_invalid(self):
-        # Given
-        mandatory_validator = validators.mandatory()
+    # When
+    lat_long_validator('1234.5678')
 
-        # When, then raises
-        with pytest.raises(validators.Invalid):
-            mandatory_validator('')
+    # Then no invalid exception is raised
 
-    def test_numeric_valid(self):
-        # Given
-        numeric_validator = validators.numeric()
 
-        # When
-        numeric_validator('0123456789')
+@pytest.mark.parametrize('value', [
+    '1.1E-10',  # Python valid scientific notation
+    '1.1e-10',
+    '2e-4',
+    '2E-4',
+    '3E4',
+    '3e5',
+    '5.9238E-7',
+    '0.2E-3',
+    '23.4E-5',
+    '1.2345E2',
+    '1.2345*10^-3',  # Not python valid notation
+    '1.2345*10**-3',
+    '1.2345x10**-3',
+    '1.2345x10^-3',
+    'eggs'
```

Comment:
Someones thinking of lunch 😆

Reply:
could we have `12.34.56.789` as a test value please... see how it handles multiple full stops